### ISSUE TITLE
__based(segment) breaks volatile semantics in C

### ISSUE THREAD

**Issue by diaphore** [open-watcom/open-watcom-v2#405]
The C compiler ignores the `volatile` qualifier when used with `__based` pointers.

```c
void based(void)
{
	__segment segment = 0x1234;
	char __based(segment) *pointer = 0;

	char dummy = *pointer;
	*pointer = dummy;
}

void volatile_based(void)
{
	__segment segment = 0x1234;
	char volatile __based(segment) *pointer = 0;

	char dummy = *pointer;
	*pointer = dummy;
}

void volatile_far(void)
{
	__segment segment = 0x1234;
	char volatile __far *pointer = segment :> (char __near *)0;

	char dummy = *pointer;
	*pointer = dummy;
}
```

In C++, both functions using volatile compile to the exact same code, as expected :
`wpp -ox test.c && wdis -l test.obj`
```
0000				void __near based():
0000  C3				ret

0001				void __near volatile_far():
0001				void __near volatile_based():
0001  53				push		bx
0002  B8 34 12				mov		ax,0x1234
0005  31 DB				xor		bx,bx
0007  8E C0				mov		es,ax
0009  26 8A 07				mov		al,byte ptr es:[bx]
000C  26 88 07				mov		byte ptr es:[bx],al
000F  5B				pop		bx
0010  C3				ret
```

In C on the other hand, the compiler seems to ignore the `volatile` qualifier entirely and thus optimize out the whole function :
`wcc -ox test.c && wdis -l test.obj`
```
0000				based_:
0000				volatile_based_:
0000  C3				ret

0001				volatile_far_:
0001  53				push		bx
0002  B8 34 12				mov		ax,0x1234
0005  31 DB				xor		bx,bx
0007  8E C0				mov		es,ax
0009  26 8A 07				mov		al,byte ptr es:[bx]
000C  26 88 07				mov		byte ptr es:[bx],al
000F  5B				pop		bx
0010  C3				ret
```

Tested on Open Watcom v2, v1.9 and old Watcom 11.c

**Comment by jmalak** [comment]
Thanks for bug report.

**Comment by jmalak** [comment]
it is fixed by c640ebc653d38ba580ab32d0f79e878501324a07 commit.

**Comment by diaphore** [comment]
It is. Thanks for your work.
